arviz/kde_utils.py

```diff
@@ -0,0 +1,895 @@
+
+
+def select_bw_method(method):
+    """
+    Selects a function to compute the bandwidth.
+    Also checks method `bw` is correctly specified.
+    Otherwise, throws an error.
+
+    Parameters
+    ----------
+    method : str
+        Method to estimate the bandwidth.
+
+    Returns
+    -------
+    bw_fun: function
+        Function to compute the bandwidth.
+    """
+
+    method_lower = method.lower()
+    if method_lower not in BW_METHODS_LINEAR.keys():
+        raise ValueError((
+            f"Unrecognized bandwidth method.\n"
+            f"Input is: {method}.\n"
+            f"Expected one of: {list(BW_METHODS_LINEAR.keys())}."
+        ))
+    bw_fun = BW_METHODS_LINEAR[method_lower]
+    return bw_fun
+
+
+def get_bw(x, bw, grid_counts=None, x_std=None, x_range=None):
+    """
+    Computes bandwidth for a given data `x` and `bw`.
+    Also checks `bw` is correctly specified.
+
+    Parameters
+    ----------
+    x : 1-D numpy array
+        1 dimensional array of sample data from the
+        variable for which a density estimate is desired.
+    bw: int, float or str
+        If numeric, indicates the bandwidth and must be positive.
+        If str, indicates the method to estimate the bandwidth.
+
+    Returns
+    -------
+    bw: float
+        Bandwidth
+    """
+    if isinstance(bw, bool):
+        raise ValueError((
+            f"`bw` must not be of type `bool`.\n"
+            f"Expected a positive numeric or one of the following strings:\n"
+            f"{list(BW_METHODS_LINEAR.keys())}."))
+    if isinstance(bw, (int, float)):
+        if bw < 0:
+            raise ValueError(f"Numeric `bw` must be positive.\nInput: {bw:.4f}.")
+    elif isinstance(bw, str):
+        bw_fun = select_bw_method(bw)
+        bw_lower = bw.lower()
+        if bw_lower == "isj":
+            bw = bw_fun(x, grid_counts, x_range)
+        elif bw_lower in ["scott", "silverman"]:
+            bw = bw_fun(x, x_std)
+        elif bw_lower == "experimental":
+            bw = bw_fun(x, grid_counts, x_std, x_range)
+    else:
+        raise ValueError((
+            f"Unrecognized `bw` argument.\n"
+            f"Expected a positive numeric or one of the following strings:\n"
+            f"{list(BW_METHODS_LINEAR.keys())}."))
+    return bw
+
+
+# Misc utils ------------------------------------------------------------------------
+def normalize_angle(x, pi_centered=False):
+    """
+    Takes angles in radians and normalize them to [-pi, pi) or [0, 2 * pi)
+    depending on `pi_centered`.
+    """
+    if pi_centered:
+        return x % (2 * np.pi)
+    else:
+        return (x + np.pi) % (2 * np.pi) - np.pi
+    
+def vonmises_pdf(x, mu, kappa):
+    assert kappa > 0, "Argument 'kappa' must be positive."
+    pdf = 1 / (2 * np.pi * ive(0, kappa)) * np.exp(np.cos(x - mu) - 1) ** kappa
+    return pdf
+
+
+def circular_mean(x):
+    sinr = np.sum(np.sin(x))
+    cosr = np.sum(np.cos(x))
+    mean = np.arctan2(sinr, cosr)
+    return mean
```

Comment:
I am aware of Scipy version but I decided to keep this one because it is faster. As it is used now, things like nan checking are done in `kde_circular()` which is the function that calls `circular_mean()`. 
I guess you mention checks and all that because you're considering using the function in other parts of arviz.
I think there are two alternatives here:

* Implement a `circular_mean()` function that is going to be used everywhere in arviz, instead of Scipy version.
* Use Scipy version in other parts of arviz and keep this function for the circular kde since it is faster and safety checks have already been performed when `circular_mean()` is called. 

What is your opinion on this?

BTW, I'm not aware if Scipy `circmean()` is currently being used in other parts of arviz, I just guess this is the default alternative.

Reply:
I think we use a circmean (not sure which) in summary with circular variables, I think we should have a single function used in both places. 

I am fine with keeping the fast one, given that it covers our use case and we have tests checking it gives the same result as scipy one.

Reply:
If you agree I can add the following function to `numeric_utils.py` and make use of it here. 
I will also add a test to check the result matches the one obtained with scipy (as far as I've manually tested, it does).

```python
def circular_mean(x, na_rm=False):   
    if len(x[np.isnan(x)]) > 0:
        if na_rm:
            x = x[~np.isnan(x)]
        else:
            return np.nan
    
    sinr = np.sum(np.sin(x))
    cosr = np.sum(np.cos(x))
    mean = np.arctan2(sinr, cosr)
    
    return mean
```

Reply:
Sounds good, I would use `np.any` instead of slicing and then checking length, and can you open an issue on using numba for this? There are other functions that could benefit from it

Reply:
what about

def circular_mean(x, na_rm=False):   
    if na_rm:
        x = x[~np.isnan(x)]
    
    sinr = np.sum(np.sin(x))
    cosr = np.sum(np.cos(x))
    mean = np.arctan2(sinr, cosr)
    
    return mean